File: backend/src/modules/charges/dto/update-charge.dto.ts
import {
  IsString,
  IsOptional,
  IsEnum,
  IsBoolean,
  IsNumber,
  IsObject,
  IsArray,
  ValidateNested,
  MaxLength,
} from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';
import { Type } from 'class-transformer';
import { BillingTiming, GraduatedRangeDto, ChargeFilterDto } from './create-charge.dto';

export class UpdateChargeDto {
  @ApiProperty({ required: false, enum: BillingTiming })
  @IsEnum(BillingTiming)
  @IsOptional()
  billingTiming?: BillingTiming;

  @ApiProperty({ required: false })
  @IsString()
  @IsOptional()
  @MaxLength(200)
  invoiceDisplayName?: string;

  @ApiProperty({ required: false })
  @IsNumber()
  @IsOptional()
  minAmountCents?: number;

  @ApiProperty({ required: false })
  @IsBoolean()
  @IsOptional()
  prorated?: boolean;

  @ApiProperty({ required: false })
  @IsObject()
  @IsOptional()
  properties?: Record<string, any>;

  @ApiProperty({ required: false, type: [GraduatedRangeDto] })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => GraduatedRangeDto)
  @IsOptional()
  graduatedRanges?: GraduatedRangeDto[];

  @ApiProperty({ required: false, type: [ChargeFilterDto] })
  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => ChargeFilterDto)
  @IsOptional()
  filters?: ChargeFilterDto[];
}

File: sdks/csharp/src/NovaBilling/Charges/Requests/UpdateChargeDto.cs
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record UpdateChargeDto
{
    /// <summary>
    /// Charge ID
    /// </summary>
    [JsonIgnore]
    public required string Id { get; set; }

    [JsonPropertyName("billingTiming")]
    public UpdateChargeDtoBillingTiming? BillingTiming { get; set; }

    [JsonPropertyName("invoiceDisplayName")]
    public string? InvoiceDisplayName { get; set; }

    [JsonPropertyName("minAmountCents")]
    public double? MinAmountCents { get; set; }

    [JsonPropertyName("prorated")]
    public bool? Prorated { get; set; }

    [JsonPropertyName("properties")]
    public Dictionary<string, object?>? Properties { get; set; }

    [JsonPropertyName("graduatedRanges")]
    public IEnumerable<GraduatedRangeDto>? GraduatedRanges { get; set; }

    [JsonPropertyName("filters")]
    public IEnumerable<ChargeFilterDto>? Filters { get; set; }

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
